Step4: Controller 메소드 인자 매핑

ghojeong opened the pull request:
개인적인 색깔을 넣기 보다는,
미션 요구사항과 Spring 구조에 집중할 수 있도록 구현해보았습니다.
jwp-jdbc 에 이미 구현되어 있는 구조를 최대한 참고하게 되었습니다.

ghojeong committed:
refactor: ModelAndView 에서 JspView 에 관한 의존성 제거

ghojeong committed:
refactor: JspView 에서 RedirectView 를 분리

ghojeong committed:
refactor: Rename declaredObject to target

ghojeong committed:
feat: 메서드의 인자를 객체화할 수 있는 MethodParameter 구현

ghojeong committed:
feat: RequestParam 어노테이션에 대한 Resolver 를 구현

ghojeong committed:
feat: Method 의 Param 에 따라, Argument 를 스캔하는 ArgumentScanner 구현

ghojeong committed:
refactor: 회원가입시, ArgumentScanner 방식에 따라 인자가 전달되도록 변경

다음 3개의 Controller 에 ArgumentScanner 가 적용되도록 함
- CreateUserController
- ForwardController
- HomeController

ghojeong committed:
refactor: ArgumentScanner 에서 조건문이 아니라, Resolver 를 통해 HttpServlet 인자를 전달하도록 수정

레거시 대응을 원래는 if 조건문으로 하였으나, 더 우아하게 수정

ghojeong committed:
feat: URL을 통해서도 동적으로 값을 전달하는 PathVariableArgumentResolver 구현

ghojeong committed:
test: PathVariableArgumentResolver 에 관한 테스트를 작성

ghojeong committed:
feat: userId Path Variable 을 통해 유저 정보를 조회하는 기능 구현

yongdae commented on src/test/java/core/mvc/resolver/RequestParamArgumentResolverTest.java:
```diff
@@ -0,0 +1,69 @@
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+class RequestParamArgumentResolverTest {
```
인자를 자동으로 맵핑해주는 기능에 대한 테스트를 상세하게 만들어주셨네요.
케이스를 요구사항에 따라 정의해주셔서 잘 정리해주셔서 쉽게 이해 할 수 있었습니다. 💯

yongdae commented on src/test/java/core/mvc/resolver/PathVariableArgumentResolverTest.java:
```diff
@@ -0,0 +1,58 @@
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+class PathVariableArgumentResolverTest {
```
👍

yongdae commented on src/main/java/core/mvc/resolver/MethodParameter.java:
```diff
@@ -0,0 +1,72 @@
+        return parameterName;
+    }
+
+    public boolean supportAnnotation(Class<?> annotationClass) {
+        return Arrays.stream(annotations).anyMatch(
+                annotation -> annotation.annotationType() == annotationClass
+        );
+    }
+
+    public <T> T getAnnotation(Class<T> annotationClass) {
+        return Arrays.stream(annotations)
+                .filter(annotation -> annotation.annotationType() == annotationClass)
+                .findAny()
+                .map(annotationClass::cast)
+                .orElseThrow(() -> new InvalidParameterAnnotation(annotationClass));
+    }
+
+    public Object resolveArgument(Object argument) {
+        if (isString()) {
+            return argument;
+        }
+        if (isInteger()) {
+            return Integer.valueOf(argument.toString());
+        }
+
+        throw new MethodArgumentTypeNotSupportedException(getType(), argument);
+    }
```
이 메서드는 PathVariableArgumentResolver, RequestParamArgumentResolver만 사용하는 부분이니
MethodParameter 대신 PathVariableArgumentResolver, RequestParamArgumentResolver가 가질 수 있게 구조를 변경해보면 좋을 것 같은데 어떻게 생각하시나요? 🤔

ghojeong replied:
어떻게 할지 고민하다가, PathVariableArgumentResolver, RequestParamArgumentResolver 의 공통 로직을 가지는
추상 클래스인 AbstractAnnotationArgumentResolver 를 선언하도록 했습니다.

HttpRequestArgumentResolver 와 HttpResponseArgumentResolver 에서는,
MethodParameter 의 위 로직이 사용되지 않으므로, 적절한 피드백 같습니다.

yongdae reviewed (changes requested):
이전 단계에 의견드린 부분을 포함해 멋지게 4단계 미션을 완수해주셨네요. 💯 
검증 뿐만 아니라 설계를 잘 해주셔서 다시 한번 더 고민하고 배울 수 있었습니다.

소소하지만 한가지 의견을 드린 부분이 있으니 확인 후 리뷰 요청 부탁드릴께요!

ghojeong committed:
refactor: PathVariableArgumentResolver, RequestParamArgumentResolver 에서만 사용하는 로직을 MethodParameter 에서 공통된 추상 클래스로 이동

MethodParameter 에서 있던 로직은 Http 관련 ArgumentResoler 에서는 사용하지 않으므로,
Annotation 과 관련된 추상 Resolver 클래스를 하나 더 만들었다.

yongdae commented on src/main/java/core/mvc/resolver/AbstractAnnotationArgumentResolver.java:
```diff
@@ -0,0 +1,34 @@
+
+import java.util.Arrays;
+
+public abstract class AbstractAnnotationArgumentResolver implements ArgumentResolver {
```
의견드린 부분을 반영해주셨네요. 👍

yongdae reviewed (approved):
의견드린 부분을 멋지게 반영해주셨네요. 💯 
다음 단계로 넘어가도록 하겠습니다. 

아 그리고 교정 수술 받으신 눈도 잘 회복되시면 좋겠네요.
수고하셨습니다. 🙇
